verilog/analysis/checkers/generate_label_prefix_rule_test.cc

```diff
@@ -0,0 +1,154 @@
+      {""},
+      {"module m;\n"
+       "endmodule\n"},
+      {"module m;\n"
+       "generate\n"
+       "if (TypeIsPosedge) ",
+       "begin",
+       "\n"
+       "  always @(posedge clk) foo <= bar;\n"
+       "end\n"
+       "endgenerate\nendmodule\n"},
+      {"module m;\n"
+       "generate\n"
+       "if (TypeIsPosedge) "
+       "begin : g_label"
+       "\n"
+       "  always @(posedge clk) foo <= bar;\n"
+       "end\n"
+       "endgenerate\nendmodule\n"},
+      {"module m;\n"
+       "generate\n"
+       "if (TypeIsPosedge) "
+       "begin : gen_label"
+       "\n"
+       "  always @(posedge clk) foo <= bar;\n"
+       "end\n"
+       "endgenerate\nendmodule\n"},
+      {"module m;\n"
+       "generate\n"
+       "if (TypeIsPosedge) "
+       "begin : gen_label"
+       "\n"
+       "  always @(posedge clk) foo <= bar;\n"
+       "end : gen_label_also\n"
+       "endgenerate\nendmodule\n"},
+      {"module m;\n"
+       "generate\n"
+       "begin\n"
+       "  initial begin : g_ini\n"
+       "    if (1) begin : gen_if\n"
+       "    end : gen_endif\n"
+       "  end : g_endini\n"
+       "end\n"
+       "endgenerate\nendmodule\n"},
+      // Test incorrect code
+      {"module m;\n"
+       "generate\n"
+       "if (TypeIsPosedge) "
+       "begin : ",
+       {SymbolIdentifier, "k_label"},
+       "\n"
+       "  always @(posedge clk) foo <= bar;\n"
+       "end\n"
+       "endgenerate\nendmodule\n"},
+      {"module m;\n"
+       "generate\n"
+       "if (TypeIsPosedge) "
+       "begin",
+       "\n"
+       "  always @(posedge clk) foo <= bar;\n"
+       "end : ",
+       {SymbolIdentifier, "genwithoutunderscore"},
+       "\nendgenerate\nendmodule\n"},
+      {"module m;\n"
+       "generate\n"
+       "if (TypeIsPosedge) ",
+       "begin : ",
+       {SymbolIdentifier, "k_label"},
+       "\n"
+       "  always @(posedge clk) foo <= bar;\n"
+       "end : ",
+       {SymbolIdentifier, "genwithoutunderscore"},
+       "\nendgenerate\nendmodule\n"},
+      // Incorrect code with more blocks
+      {"module m;\n"
+       "generate\n"
+       "begin : ",
+       {SymbolIdentifier, "wrong"},
+       "\n"
+       "  initial begin : g_ini\n"
+       "    if (1) begin : gen_if\n"
+       "    end : gen_endif\n"
+       "  end : g_endini\n"
+       "end\n"
+       "endgenerate\nendmodule\n"},
+
+      {"module m;\n"
+       "generate\n"
+       "begin\n",
+       "  initial begin : ",
+       {SymbolIdentifier, "wrong"},
+       "\n"
+       "    if (1) begin : gen_if\n"
+       "    end : gen_endif\n"
+       "  end : g_endini\n"
+       "end\n"
+       "endgenerate\nendmodule\n"},
+
+      {"module m;\n"
+       "generate\n"
+       "begin\n",
+       "  initial begin : ",
+       {SymbolIdentifier, "wrong"},
+       "\n"
+       "    if (1) begin : gen_if\n"
+       "    end : ",
+       {SymbolIdentifier, "also_wrong"},
+       "\n"
+       "  end : g_endini\n"
+       "end\n"
+       "endgenerate\nendmodule\n"},
+  };
```

Comment:
Suggested test cases:

* loop generate statements (not inside `generate` and `endgenerate`), so `module m; if (x) ...`
  * with and without `begin` `end` labels
* conditional generate statements (not inside `generate` and `endgenerate`), so `module ; for(...`
  * with and without `begin` `end` labels
* case generate statement (doesn't have `begin` `end`, so a negative test)
* empty generate region (doesn't have `begin` `end`, so a negative test)
* always/initial/final constructs with begin/end labels at the module-item level (not inside generate/endgenerate), should be negative test

Reply:
This case is still missing (and is considered a generate label):

```systemverilog
module m;
if (x) begin : bad_label
end : bad_label
endmodule
```


Negative test:

```systemverilog
module m;
initial begin : OkNotAGenerateLabel
end : OkNotAGenerateLabel
endmodule
```